spinoso-string/src/iter.rs

```diff
@@ -0,0 +1,589 @@
+/// [Ruby `ArgumentError` Exception class]: https://ruby-doc.org/core-2.6.3/ArgumentError.html
+/// [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum CenterError {
+    /// Error returned when calling [`String::center`] with an empty padding
+    /// byte string.
+    ///
+    /// [`String::center`]: crate::String::center
+    ZeroWidthPadding,
+}
+
+impl CenterError {
+    pub const EXCEPTION_TYPE: &'static str = "ArgumentError";
+
+    /// Create a new zero width padding `CenterError`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use spinoso_string::CenterError;
+    ///
+    /// const ERR: CenterError = CenterError::zero_width_padding();
+    /// assert_eq!(ERR.message(), "zero width padding");
+    /// ```
+    #[inline]
+    #[must_use]
+    pub const fn zero_width_padding() -> Self {
+        Self::ZeroWidthPadding
+    }
+
+    /// Retrieve the exception message associated with this center error.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use spinoso_string::CenterError;
+    /// let err = CenterError::zero_width_padding();
+    /// assert_eq!(err.message(), "zero width padding");
+    /// ```
+    #[inline]
+    #[must_use]
+    #[allow(clippy::unused_self)]
+    pub const fn message(self) -> &'static str {
+        "zero width padding"
+    }
+}
+
+impl fmt::Display for CenterError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let CenterError::ZeroWidthPadding = self;
+        f.write_str(self.message())
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for CenterError {}
+
+/// An iterator that yields a byte string centered within a padding byte string.
+///
+/// This struct is created by the [`center`] method on a Spinoso [`String`]. See
+/// its documentation for more.
+///
+/// # Examples
+///
+/// ```
+/// use spinoso_string::String;
+/// # fn example() -> Result<(), spinoso_string::CenterError> {
+/// let s = String::from("hello");
+///
+/// assert_eq!(s.center(4, None)?.collect::<Vec<_>>(), b"hello");
+/// assert_eq!(s.center(20, None)?.collect::<Vec<_>>(), b"       hello        ");
+/// assert_eq!(s.center(20, Some(&b"123"[..]))?.collect::<Vec<_>>(), b"1231231hello12312312");
+/// # Ok(())
+/// # }
+/// # example().unwrap();
+/// ```
+///
+/// This iterator is [encoding-aware]. [Conventionally UTF-8] strings are
+/// iterated by UTF-8 byte sequences.
+///
+/// ```
+/// use spinoso_string::String;
+/// # fn example() -> Result<(), spinoso_string::CenterError> {
+/// let s = String::from("💎");
+///
+/// assert_eq!(s.center(3, None)?.collect::<Vec<_>>(), " 💎 ".as_bytes());
+/// # Ok(())
+/// # }
+/// # example().unwrap();
+/// ```
+///
+/// [`String`]: crate::String
+/// [`center`]: crate::String::center
+/// [encoding-aware]: crate::Encoding
+/// [Conventionally UTF-8]: crate::Encoding::Utf8
+#[derive(Debug, Clone)]
+pub struct Center<'a, 'b> {
```

Comment:
`Center` is a bit more complex of an iterator (with its error type and all). Let's move `Center` and its associated impls and types to a new `center.rs` file, like how `Chars` has its own file.